### PR TITLE
Add a base workspace template with peering

### PR DESCRIPTION
**Pull request by deniscep**
# Add a base workspace template with peering

## What is being addressed

Adds a base workspace template that can be used for creation of other workspace templates. Doesn't contain any services, only VNET with one subnet and the VNET is peered.

## How is this addressed

- Add a basic terraform structure for workspace creation
- Add a workspace resource group with VNET, subnet and peering to Core VNET.

Closes 5484 in AzDO

### CODE REVIEW

File: templates/workspaces/base_workspace/terraform/locals.tf [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+data "azurerm_client_config" "current" {}
+
+# Random unique id
+resource "random_string" "unique_id" {
+  length      = 4
+  min_numeric = 4
+}
+
+locals {
+  tre_id                                = random_string.unique_id.result
```
Comment (marrobi) [inline templates/workspaces/base_workspace/terraform/locals.tf:12]: Should this be workspace_id ?

File: templates/workspaces/base_workspace/terraform/network/locals.tf [hunk 2]
```diff
@@ -0,0 +1,4 @@
+locals {
+  ws_services_vnet_subnets            = cidrsubnets(var.address_space, 4)
```
Comment (marrobi) [inline templates/workspaces/base_workspace/terraform/network/locals.tf:2]: @christoferlof think we need another issue to properly plan workspace subnets.

File: templates/workspaces/base_workspace/terraform/network/network.tf [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+  address_space       = [var.address_space]
+}
+
+
+resource "azurerm_subnet" "services" {
+  name                                           = "ServicesSubnet"
+  virtual_network_name                           = azurerm_virtual_network.ws.name
+  resource_group_name                            = var.resource_group_name
+  address_prefixes                               = [local.services_subnet_address_prefix]
+  enforce_private_link_endpoint_network_policies = true
```
Comment (marrobi) [inline templates/workspaces/base_workspace/terraform/network/network.tf:14]: Can we add a comment here maybe to highlight that this settings means private endpoints do not adhere to NSG rules applied to them? I believe the TF provider has the incorrect naming which causes confusion.

**Review by marrobi — COMMENTED** [review]
@deniscep Couple of comments - not all for yourself... Based on the infra proposal there are a number of items missing fromt he base workspace template. Have we got issues tracking these?

- NSGs
- App GW
- Storage
- Route Table

Thanks.

**Comment by christoferlof** [comment]
#53 and #51 are related

File: templates/workspaces/base_workspace/terraform/locals.tf (hunk 1 above)
Comment (deniscep) [reply templates/workspaces/base_workspace/terraform/locals.tf:12]: Good point, id should probably be sequential; I'll change.

File: templates/workspaces/base_workspace/terraform/main.tf [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+  }
+}
+
+provider "azurerm" {
+    features {}
+}
+
+resource "azurerm_resource_group" "ws" {
+  location  = var.location
+  name      = "rg-ws-${var.resource_name_prefix}-${var.environment}-${local.tre_id}"
```
[the remark this answers is not in the document]
Comment (deniscep) [reply templates/workspaces/base_workspace/terraform/main.tf:17]: Good point, I'll change the naming convention for workspace to be <ResoruceAbrv>-<CoreTREENVName>-ws-<seqID>

File: templates/workspaces/base_workspace/terraform/terraform.tfvars.tmpl [hunk 5]
```diff
@@ -0,0 +1,6 @@
+resource_name_prefix    = "tre"
+environment             = "dev"
+location                = "westeurope"
+address_space           = "10.1.1.0/24"
+core_vnet               = "dc55trevnet"
+core_resource_group_name= "dc55trerg"
```
Comment (TessFerrandez) [inline templates/workspaces/base_workspace/terraform/terraform.tfvars.tmpl:6]:
```suggestion
resource_name_prefix    = "tre"
environment             = "dev"
location                = "westeurope"
address_space           = "10.1.1.0/24"
core_vnet               = "dc56trevnet"
core_resource_group_name= "dc55trerg"
```

Comment (TessFerrandez) [inline templates/workspaces/base_workspace/terraform/terraform.tfvars.tmpl:6]:
```suggestion
resource_name_prefix    = "tre"
environment             = "dev"
location                = "westeurope"
address_space           = "10.1.1.0/24"
core_vnet               = "dc56trevnet"
core_resource_group_name= "dc55trerg"
```

**Review by TessFerrandez — APPROVED** [review]

File: templates/workspaces/base_workspace/terraform/network/network.tf (hunk 3 above)
Comment (deniscep) [reply templates/workspaces/base_workspace/terraform/network/network.tf:14]: added

**Pull request merged**